src/v/security/audit/schemas/types.h

```diff
@@ -0,0 +1,478 @@
+    w.EndObject();
+}
+
+inline void
+rjson_serialize(Writer<StringBuffer>& w, const sa::resource_detail& resource) {
+    w.StartObject();
+    w.Key("name");
+    rjson_serialize(w, resource.name);
+    w.Key("type");
+    rjson_serialize(w, resource.type);
+    w.EndObject();
+}
+
+inline void
+rjson_serialize(Writer<StringBuffer>& w, const sa::authorization_metadata& m) {
+    w.StartObject();
+    w.Key("acl_authorization");
+    w.StartObject();
+    w.Key("host");
+    ::json::rjson_serialize(w, m.acl_authorization.host);
+    w.Key("op");
+    ::json::rjson_serialize(w, m.acl_authorization.op);
+    w.Key("permission_type");
+    ::json::rjson_serialize(w, m.acl_authorization.permission_type);
+    w.Key("principal");
+    ::json::rjson_serialize(w, m.acl_authorization.principal);
+    w.EndObject();
+    w.Key("resource");
+    w.StartObject();
+    w.Key("name");
+    ::json::rjson_serialize(w, m.resource.name);
+    w.Key("pattern");
+    ::json::rjson_serialize(w, m.resource.pattern);
+    w.Key("type");
+    ::json::rjson_serialize(w, m.resource.type);
+    w.EndObject();
+    w.EndObject();
+}
+
+inline void
+rjson_serialize(Writer<StringBuffer>& w, const sa::api_activity_unmapped& u) {
+    w.StartObject();
+    w.Key("shard_id");
+    ::json::rjson_serialize(w, u.shard_id);
+    if (u.authorization_metadata) {
+        w.Key("authorization_metadata");
+        ::json::rjson_serialize(w, u.authorization_metadata.value());
+    }
+    w.EndObject();
+}
+
+inline void rjson_serialize(Writer<StringBuffer>& w, const sa::http_header& h) {
+    w.StartObject();
+    w.Key("name");
+    json::rjson_serialize(w, h.name);
+    w.Key("value");
+    json::rjson_serialize(w, h.value);
+    w.EndObject();
+}
+
+inline void rjson_serialize(
+  Writer<StringBuffer>& w, const sa::uniform_resource_locator& url) {
+    w.StartObject();
+    w.Key("hostname");
+    rjson_serialize(w, url.hostname);
+    w.Key("path");
+    rjson_serialize(w, url.path);
+    w.Key("port");
+    rjson_serialize(w, url.port);
+    w.Key("scheme");
+    rjson_serialize(w, url.scheme);
+    w.Key("url_string");
+    rjson_serialize(w, url.url_string);
+    w.EndObject();
+}
+
+inline void
+rjson_serialize(Writer<StringBuffer>& w, const sa::http_request& r) {
+    w.StartObject();
+    w.Key("http_headers");
+    rjson_serialize(w, r.http_headers);
+    w.Key("http_method");
+    rjson_serialize(w, r.http_method);
+    w.Key("url");
+    rjson_serialize(w, r.url);
+    w.Key("user_agent");
+    rjson_serialize(w, r.user_agent);
+    w.Key("version");
+    rjson_serialize(w, r.version);
+    w.EndObject();
+}
+} // namespace json
+
+namespace std {
+
+namespace sa = security::audit;
+
+template<typename T>
+struct hash<std::vector<T>> {
+    size_t operator()(const std::vector<T>& v) const {
+        size_t h = 0;
+        for (const auto& val : v) {
+            boost::hash_combine(
+              h, std::hash<std::remove_cvref_t<decltype(val)>>()(val));
```

Comment:
can you write `std::hash<T>` instead of remove_cvref...

Reply:
yup that works